### PR TITLE
JVM-2081: PEA materializes the object at bytecode aastore

### PR DESCRIPTION
**Pull request by navyxliu**
This patch intercepts bytecode aastore in C2 Parse.

It checks both array and object. If the array is non-tracking or has been Escaped and object is tracking and still Virtual, PEA materializes object.  This is the case of Blender.java, which is Sunflow's kernel. 

here are some quick data. we can see that C2 with PEA performs better than C2 but not as good as Graal JIT yet. We will take a closer look at this test and Dacapo/Sunflow further.  

```
// C2 PEA 
➜  PEA git:(JVM_2055) ../build/linux-x86_64-server-release/images/jdk/bin/java -XX:+UnlockExperimentalVMOptions -XX:+DoPartialEscapeAnalysis Blender
1320 ms
1193 ms
1211 ms
1189 ms
1221 ms
1234 ms
1240 ms
1236 ms
1234 ms
1234 ms
// C2 without PEA(default)
➜  PEA git:(JVM_2055) ../build/linux-x86_64-server-release/images/jdk/bin/java -XX:+UnlockExperimentalVMOptions -XX:-DoPartialEscapeAnalysis Blender
2350 ms
1745 ms
1737 ms
1729 ms
1795 ms
1800 ms
1789 ms
1791 ms
1782 ms
1788 ms
// Graal JIT with PEA(default)
➜  compiler git:(46f8f602aca) mx vm -cp ~/Devel/jdk/PEA Blender
2238 ms
1012 ms
1009 ms
1012 ms
1014 ms
1009 ms
1008 ms
1033 ms
1026 ms
1007 ms
// Greal JIT without PEA
➜  compiler git:(46f8f602aca) mx vm -cp ~/Devel/jdk/PEA -Dgraal.PartialEscapeAnalysis=false Blender
3334 ms
2109 ms
1820 ms
1671 ms
1665 ms
1897 ms
1743 ms
1744 ms
1743 ms
1733 ms
```

### CODE REVIEW

File: src/hotspot/share/opto/parse2.cpp [hunk 1]
```diff
@@ -99,6 +99,17 @@ void Parse::array_store(BasicType bt) {
...
     bt = T_BOOLEAN;
   }
+
+  if (DoPartialEscapeAnalysis && bt == T_OBJECT) {
+    PEAState& as = jvms()->alloc_state();
+    VirtualState* obj_vs = as.as_virtual(val);    // obj is tracked and Virtual
+    VirtualState* ary_vs = as.as_virtual(array);  // array is not-tracked or Escaped
+
+    if (ary_vs == nullptr && obj_vs != nullptr) {
+      EscapedState* escaped = as.materialize(this, val);
```
Comment (caojoshua) [inline src/hotspot/share/opto/parse2.cpp:109]: Can we consider this a bug fix? I'm surprised we have not run into this before. I guess raw array's are not used so much in Java code.

**Review by caojoshua — APPROVED** [review]

File: src/hotspot/share/opto/parse2.cpp (hunk 1 above)
Comment (navyxliu) [reply src/hotspot/share/opto/parse2.cpp:109]: yes, it's a bug. Prior to this patch,  we convert an object from Virtual to Escaped in the following 2 cases.
1. put-field and put-static-field (GlobalEscaped)
2. invoke bytecodes (ArgEscaped)

This is the 3rd case. Are there anymore cases?  
3. aatore (GlobalEscaped)

We also convert an object from Virtual to Escaped when we encounter a loop header, but we don't materialize it.

Comment (navyxliu) [reply src/hotspot/share/opto/parse2.cpp:109]: @caojoshua one thing I don't quite understand here. Is it possible that `array` is null here? 

`array_store_check()` before performs subtype check, but I don't find explicit check that array is null. it feel that C2 assume it's impossible. do you know why?

Comment (caojoshua) [reply src/hotspot/share/opto/parse2.cpp:109]: Yeah, I don't see a explicit check. My guess is that, if the input Java bytecode is correct, C2 can assume that there is an array pointer on the stack. You could ask the same question about `val`.

There are many instances of `pop()`, and I'm not seeing much null checks.

**Pull request merged**